File: 07-06-2024_Replace_Words.py
class Trie:
    def __init__(self):
        self.children = [None] * 26
        self.is_end = False

    def insert(self, word, pos=0):
        if pos < len(word):
            index = ord(word[pos]) - ord('a')
            if not self.children[index]:
                self.children[index] = Trie()
            if self.children[index].is_end:
                return
            self.children[index].insert(word, pos + 1)
        else:
            self.is_end = True

    def find(self, word, pos=0):
        if pos == len(word):
            return word, self.is_end
        index = ord(word[pos]) - ord('a')
        if self.children[index]:
            if self.children[index].is_end:
                return word[:pos + 1], True
            return self.children[index].find(word, pos + 1)
        return word, False


class Solution:
    def replaceWords(self, dictionary, sentence):
        trie = Trie()
        for root in dictionary:
            trie.insert(root)
        
        result = []
        words = sentence.split()
        for word in words:
            replacement, _ = trie.find(word)
            result.append(replacement)
        
        return ' '.join(result)

File: 07-06-2024_Replace_Words.java
import java.util.List;
import java.util.StringJoiner;

class Solution {
    public static class Trie {
        private Trie[] children = new Trie[26];
        private boolean isEnd;

        public void insert(String word) {
            Trie node = this;
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.children[index] == null) {
                    node.children[index] = new Trie();
                }
                node = node.children[index];
            }
            node.isEnd = true;
        }

        public String findRoot(String word) {
            Trie node = this;
            StringBuilder prefix = new StringBuilder();
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.children[index] == null) {
                    break;
                }
                node = node.children[index];
                prefix.append(c);
                if (node.isEnd) {
                    return prefix.toString();
                }
            }
            return word;
        }
    }

    public String replaceWords(List<String> dictionary, String sentence) {
        Trie trie = new Trie();
        for (String root : dictionary) {
            trie.insert(root);
        }
        
        StringJoiner result = new StringJoiner(" ");
        String[] words = sentence.split(" ");
        for (String word : words) {
            result.add(trie.findRoot(word));
        }
        
        return result.toString();
    }
}
